pallets/asset-depository/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+        Twox64Concat,
+        T::AssetId,
+        AccountBalance<T::Balance>,
+        ValueQuery,
+    >;
+
+    /// The aggregated balance of an asset.
+    #[pallet::storage]
+    pub type TotalBalance<T: Config> =
+        StorageMap<_, Twox64Concat, T::AssetId, T::Balance, ValueQuery>;
+
+    #[pallet::event]
+    pub enum Event<T: Config> {}
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Thrown when depositing an amount of an asset has caused an overflow of the aggregated balance.
+        TotalBalanceOverflow,
+        /// Thrown when depositing amount into an user account caused an overflow.
+        BalanceOverflow,
+        /// Thrown when withdrawing would cause an underflow
+        NotEnoughBalance,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {}
+
+    impl<T: Config> Pallet<T> {
+        /// Set the available balance of the given account the given value.
+        pub(crate) fn set_available_balance(
+            asset_id: &T::AssetId,
+            who: &AccountIdFor<T>,
+            amount: T::Balance,
+        ) {
+            Accounts::<T>::mutate(who, asset_id, |account| {
+                account.available = amount;
+            });
+        }
+    }
+
+    impl<T: Config> MultiAssetDepository<T::AssetId, AccountIdFor<T>, T::Balance> for Pallet<T> {
+        /// The total amount of the given asset currently held
+        fn aggregated_balance(asset_id: &T::AssetId) -> T::Balance {
+            TotalBalance::<T>::get(asset_id)
+        }
+
+        /// The total balance of an asset of a user
+        fn total_balance(asset_id: &T::AssetId, who: &AccountIdFor<T>) -> T::Balance {
+            Accounts::<T>::get(who, asset_id).total_balance()
+        }
+
+        /// The current available balance of an asset of a user
+        fn available_balance(asset_id: &T::AssetId, who: &AccountIdFor<T>) -> T::Balance {
+            Accounts::<T>::get(who, asset_id).available
+        }
+
+        /// Deposit the `amount` of the given asset into the available balance of the given account `who`.
+        fn deposit(
+            asset_id: &T::AssetId,
+            who: &AccountIdFor<T>,
+            amount: T::Balance,
+        ) -> DispatchResult {
+            if amount.is_zero() {
+                return Ok(());
+            }
+
+            TotalBalance::<T>::try_mutate(asset_id, |total| -> DispatchResult {
+                *total = total
+                    .checked_add(&amount)
+                    .ok_or(Error::<T>::TotalBalanceOverflow)?;
+
+                // SAFETY: this can't overflow because the balance for an account is
+                //  at most equal to the total balance and adding to it already succeeded
+                Self::set_available_balance(
+                    &asset_id,
+                    who,
+                    Self::available_balance(&asset_id, who) + amount,
+                );
```

Comment:
Why not using https://substrate.dev/rustdocs/v3.0.0/sp_arithmetic/traits/trait.Saturating.html#tymethod.saturating_add ?

Reply:
if we were to use saturating, we would be guaranteed no error, but the information would be lost as to how much we are over, which would cause us to leak balance. This should definitely end in an error if we do not have enough more capacity to hold higher values.